File: leetcode-cn/54.螺旋矩阵.py
#
# @lc app=leetcode.cn id=54 lang=python3
#
# [54] 螺旋矩阵
#

# @lc code=start
from typing import List


class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        m = len(matrix)
        if m == 0:
            return []
        n = len(matrix[0])

        ans = []
        top = 0
        buttom = m - 1
        left = 0
        right = n - 1
        while (left <= right and top <= buttom):
            for i in range(left, right + 1):
                ans.append(matrix[top][i])
            for i in range(top + 1, buttom + 1):
                ans.append(matrix[i][right])

            if left < right and top < buttom:
                for i in range(right - 1, left, -1):
                    ans.append(matrix[buttom][i])
                for i in range(buttom, top, -1):
                    ans.append(matrix[i][left])

            left += 1
            right -= 1
            top += 1
            buttom -= 1
        return ans


# @lc code=end

File: leetcode-cn/C++/54.螺旋矩阵.cpp
/*
 * @lc app=leetcode.cn id=54 lang=cpp
 *
 * [54] 螺旋矩阵
 */

// @lc code=start
#include <vector>
using namespace std;

class Solution
{
public:
    vector<int> spiralOrder(vector<vector<int>> &matrix)
    {
        int m = matrix.size();
        if (m == 0)
        {
            return {};
        }
        int n = matrix[0].size();
        int top = 0;
        int bottom = m - 1;
        int left = 0;
        int right = n - 1;

        vector<int> ans;

        while (left <= right && top <= bottom)
        {
            for (int i = left; i <= right; i++)
            {
                ans.push_back(matrix[top][i]);
            }
            top++;

            for (int i = top; i <= bottom; i++)
            {
                ans.push_back(matrix[i][right]);
            }
            right--;

            if (left <= right && top <= bottom)
            {
                for (int i = right; i >= left; i--)
                {
                    ans.push_back(matrix[bottom][i]);
                }
                bottom--;

                for (int i = bottom; i >= top; i--)
                {
                    ans.push_back(matrix[i][left]);
                }
                left++;
            }
        }

        return ans;
    }
};
// @lc code=end
